src/vos/vos_internal.h

```diff
@@ -146,10 +160,73 @@ struct vos_container {
 	struct vea_hint_context	*vc_hint_ctxt[VOS_IOS_CNT];
 	/* Various flags */
 	unsigned int		vc_in_aggregation:1,
-				vc_abort_aggregation:1;
+				vc_abort_aggregation:1,
+				vc_reindex_cmt_dtx:1;
 	unsigned int		vc_open_count;
+	uint64_t		vc_dtx_resync_gen;
+
+	/* The list of dtx_batched_cleanup_blob::bcb_cont_link. */
+	d_list_t		vc_batched_cleanup_list;
 };
 
+struct dtx_batched_cleanup_blob {
+	/* Link into vos_container::vc_batched_cleanup_list */
+	d_list_t		bcb_cont_link;
+
+	/* The list of vos_dtx_cmt_ent::dce_bcb_link. */
+	d_list_t		bcb_dce_list;
+
+	/* The count of valid DAE entry, if hit zero, then batched cleanup. */
+	int			bcb_dae_count;
+
+	/* The offset of the vos_dtx_scm_blob to be batched cleanup. */
+	umem_off_t		bcb_dsb_off;
+
+	/* The offset of dae_rec_off to be batched cleanup. */
+	umem_off_t		bcb_recs[0];
+};
+
+struct vos_dtx_act_ent {
+	struct vos_dtx_act_ent_df	 dae_base;
+	umem_off_t			 dae_df_off;
+	struct vos_dtx_scm_blob		*dae_dsb;
+	struct dtx_batched_cleanup_blob	*dae_bcb;
+	/* More DTX records if out of the inlined buffer. */
+	struct vos_dtx_record_df	*dae_records;
+	/* The capacity of dae_records, NOT including the inlined buffer. */
+	int				 dae_rec_cap;
+};
+
+#define DAE_XID(dae)		((dae)->dae_base.dae_xid)
+#define DAE_OID(dae)		((dae)->dae_base.dae_oid)
+#define DAE_DKEY_HASH(dae)	((dae)->dae_base.dae_dkey_hash)
+#define DAE_EPOCH(dae)		((dae)->dae_base.dae_epoch)
+#define DAE_SRV_GEN(dae)	((dae)->dae_base.dae_srv_gen)
+#define DAE_LAYOUT_GEN(dae)	((dae)->dae_base.dae_layout_gen)
+#define DAE_INTENT(dae)		((dae)->dae_base.dae_intent)
+#define DAE_INDEX(dae)		((dae)->dae_base.dae_index)
+#define DAE_REC_INLINE(dae)	((dae)->dae_base.dae_rec_inline)
+#define DAE_FLAGS(dae)		((dae)->dae_base.dae_flags)
+#define DAE_REC_CNT(dae)	((dae)->dae_base.dae_rec_cnt)
+#define DAE_REC_OFF(dae)	((dae)->dae_base.dae_rec_off)
```

Comment:
this macros makes the code hard to read, I think whenever we want to use them, we could just add local variable in the function which points to dae::dae_base.

Reply:
The macro in the code helps to distinguish whether is for DRAM data/modification or in SCM. If we use local variable to pointer dae::dae_base, we may need some special name.